### PR TITLE
Add option to create GitHub Deployment (environment-matrix)

### PR DESCRIPTION
**Pull request by int128**
## New feature
This will add an option to create GitHub Deployment in environment-matrix action.
It is useful to receive the deployment status from an external system such as Argo CD.
See README for details.

### CODE REVIEW

File: environment-matrix/src/matcher.ts [hunk 1]
```diff
@@ -15,42 +15,14 @@ export const find = (context: Context, rules: Rules): Environment[] | undefined
...
-  assert(x.base != null)
-  assert('ref' in x.base)
-  assert(typeof x.base.ref === 'string')
-
-  assert('head' in x)
-  assert(typeof x.head === 'object')
-  assert(x.head != null)
-  assert('ref' in x.head)
-  assert(typeof x.head.ref === 'string')
-}
```
Comment (int128) [inline environment-matrix/src/matcher.ts:56]: Moved to `environment-matrix/src/github.ts`.

File: environment-matrix/src/github.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+  assert(x.base != null)
+  assert('ref' in x.base)
+  assert(typeof x.base.ref === 'string')
+
+  assert('head' in x)
+  assert(typeof x.head === 'object')
+  assert(x.head != null)
+  assert('ref' in x.head)
+  assert(typeof x.head.ref === 'string')
+}
```
Comment (int128) [inline environment-matrix/src/github.ts:37]: Moved from `environment-matrix/src/matcher.ts`.

File: environment-matrix/src/deployment.ts [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+  for (const environment of environments) {
+    const { overlay, namespace } = environment
+    if (overlay && namespace && service) {
+      const deployment = await createDeployment(octokit, context, overlay, namespace, service)
+      environment['github-deployment-url'] = deployment.url
+    }
+  }
+}
+
+const createDeployment = async (
```
Comment (int128) [inline environment-matrix/src/deployment.ts:25]: Mostly same as https://github.com/int128/deployment-action/blob/ebb1d7f743c85e5ab8cc1b597c0664b9997c0dee/src/run.ts#L19-L56

File: environment-matrix/src/run.ts [hunk 4]
```diff
@@ -2,25 +2,40 @@ import * as core from '@actions/core'
...
     throw new Error(`no environment to deploy`)
   }
-  core.info(`environments: ${JSON.stringify(environments, undefined, 2)}`)
-  return {
-    environments,
+
+  core.info(`environments = ${JSON.stringify(environments, undefined, 2)}`)
+  if (!inputs.service) {
+    return { environments }
   }
```
Comment (panicboat) [inline environment-matrix/src/run.ts:29]: 📝 service が指定されなかった場合は挙動に変更なし

File: environment-matrix/src/matcher.ts [hunk 5]
```diff
@@ -15,42 +15,14 @@ export const find = (context: Context, rules: Rules): Environment[] | undefined
...
 const match = (context: Context, rule: Rule): boolean => {
   if (context.eventName === 'pull_request' && rule.pull_request !== undefined) {
-    const { pull_request } = context.payload
-    assertPullRequestPayload(pull_request)
+    assertPullRequestPayload(context.payload.pull_request)
     return (
-      minimatch(pull_request.base.ref, rule.pull_request.base) &&
-      minimatch(pull_request.head.ref, rule.pull_request.head)
+      minimatch(context.payload.pull_request.base.ref, rule.pull_request.base) &&
+      minimatch(context.payload.pull_request.head.ref, rule.pull_request.head)
```
Comment (panicboat) [inline environment-matrix/src/matcher.ts:21]: 📝 context
```
Pick<typeof github.context, 'eventName' | 'ref' | 'payload'>
```

**Review by panicboat — APPROVED** [review]

File: environment-matrix/src/matcher.ts (hunk 5 above)
Comment (semnil) [reply environment-matrix/src/matcher.ts:21]: https://github.com/quipper/monorepo-deploy-actions/blob/49c0eba14dbea7adfc546c1bb182d1f54f204f43/environment-matrix/src/deployment.ts#L8

**Review by semnil — APPROVED** [review]

**Pull request merged**